llama-cpp-2/src/token/data_array.rs

```diff
@@ -47,4 +57,320 @@ impl LlamaTokenDataArray {
     {
         Self::new(data.into_iter().collect(), sorted)
     }
+
+    #[must_use]
+    pub fn selected_token(&self) -> Option<LlamaToken> {
+        self.data
+            .get(usize::try_from(self.selected).ok()?)
+            .map(LlamaTokenData::id)
+    }
+}
+
+impl LlamaTokenDataArray {
+    /// Modify the underlying data as a `llama_token_data_array`. and reconstruct the `LlamaTokenDataArray`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if some of the safety conditions are not met. (we cannot check all of them at runtime so breaking them is UB)
+    ///
+    /// SAFETY:
+    /// [modify] cannot change the data pointer.
+    /// if the data is not sorted, sorted must be false.
+    /// the size of the data can only decrease (i.e you cannot add new elements).
+    pub(crate) unsafe fn modify_as_c_llama_token_data_array<T>(
+        &mut self,
+        modify: impl FnOnce(&mut llama_cpp_sys_2::llama_token_data_array) -> T,
+    ) -> T {
+        let size = self.data.len();
+        let data = self.data.as_mut_ptr().cast();
```

Comment:
unless there are platform considerations, I think explicitly mentioning the type to cast to would be best here.

Reply:
Done!